// Dobin/lib/Lib/Util/Mysql.js
var mysql = require("mysql"),
	thunkify = require('thunkify-wrap'),
	fs = require("fs")
	;

//声明mysql类
var Mysql = function(options){
	this.con =  mysql.createConnection($.extend($.Config("mysql"),options));
	this._query = thunkify(this.con.query,this.con);
}

Mysql.prototype = {
	query : function *( sql ) {
		var rows = yield this._query( sql );
		this._log(sql);
		return  rows;
	},
	getAll : function *( sql ) {
		var rows = yield this._query( sql );
		this._log(sql);
		return  rows[0];
	},
	getOne : function *( sql ) {
		var rows = yield this._query( sql );
		this._log(sql);
		return  rows[0][0];
	},
	//返回插入ID
	add : function *( sql ) {
		var rows = yield this._query( sql );
		this._log(sql);
		return  rows[0]['insertId'] || 0;
	},
	update : function *( sql ) {
		var rows = yield this._query( sql );
		this._log(sql);
		return  rows[0]['affectedRows'] || 0;
	},
	_log : function ( data ) {
		if( !fs.existsSync($.DOBIN_PATH+"/Log") ){
			fs.mkdirSync($.DOBIN_PATH+"/Log");
		}
		if( data === null ){
			fs.writeFile($.DOBIN_PATH+"/Log/mysql.log",'',function (err) {
				if(err){
					console.log(err);
				}
			});
		}
		fs.appendFile($.DOBIN_PATH+"/Log/mysql.log",data + '\n',function (err) {
			if(err){
				console.log(err);
			}
		});
	},
	close :function () {
		this.con.end();
	}
};


//挂载到模块
module.exports = Mysql;


//外部全局接口
$.mysql_query = function ( sql, options ){
	return function *(){
		var mysql = new (require($.DOBIN_PATH+"/Lib/Util/Mysql.js"))(options);
		var res = yield mysql.query( sql );
		mysql.close();
		return res;
	};
}
$.mysql_getAll = function ( sql, options ){
	return function *(){
		var mysql = new (require($.DOBIN_PATH+"/Lib/Util/Mysql.js"))(options);
		var res = yield mysql.getAll( sql );
		mysql.close();
		return res;
	};
}
$.mysql_getOne = function ( sql, options ){
	return function *(){
		var mysql = new (require($.DOBIN_PATH+"/Lib/Util/Mysql.js"))(options);
		var res = yield mysql.getOne( sql );
		mysql.close();
		return res;
	};
}
$.mysql_add = function ( sql, options ){
	return function *(){
		var mysql = new (require($.DOBIN_PATH+"/Lib/Util/Mysql.js"))(options);
		var res = yield mysql.add( sql );
		mysql.close();
		return res;
	};
}
$.mysql_update = function ( sql, options ){
	return function *(){
		var mysql = new (require($.DOBIN_PATH+"/Lib/Util/Mysql.js"))(options);
		var res = yield mysql.update( sql );
		mysql.close();
		return res;
	};
}
