// Dobin/lib/Lib/Base/Server.js
//Server.js 用于生成APP运行的上下文

'use strict';


var path = require('path'),
	fs = require('fs'),
	events = require('events')
	;

module.exports = {
	run: function () {
		this.init();

		//加载Util库
		this.loadUtil();

		//载入配置文件
		this.loadConfig();

		//触发事件
		$.events.emit("server_init");

		//写入日志
		$.server_log('Server started at ' + new Date());
	},

	//加载基础框架
	init : function  () {
		//设置基本路径
		var config = {
			LOG_PATH: $.DOBIN_PATH + '/Log',
			COMMON_PATH: $.APP_PATH + '/Common',
			CONF_PATH: $.APP_PATH + '/Conf',
			LANG_PATH: $.APP_PATH + '/Lang',
			VIEW_PATH: $.APP_PATH + '/View',
		};
		for (var name in config) {
			if ($[name] === undefined) {
				$[name] = config[name];
			}
		}

		//加载基础函数库
		require($.DOBIN_PATH + "/Lib/Common/function.js");

		//加载事件对象
		$.events = new events.EventEmitter();
	},

	//加载Util库
	loadUtil : function () {
		var sys_common = require($.DOBIN_PATH + "/Lib/Common/common.js"),
			user_common = require($.COMMON_PATH + "/common.js"),
			utils = sys_common.utils.concat(user_common.utils)
			;
		utils.forEach(function (item) {
			if( fs.existsSync($.DOBIN_PATH+"/Lib/Util/"+item+".js") )
				require($.DOBIN_PATH+"/Lib/Util/"+item+".js");
		});
	},

	//加载配置项
	loadConfig : function  () {
		//加载系统配置项
		$.Config(require($.DOBIN_PATH + "/Lib/Conf/config.js"));
		//加载用户配置项
		if(fs.existsSync($.CONF_PATH + "/config.js"))
			$.Config($.CONF_PATH + "/config.js");
	}
}
